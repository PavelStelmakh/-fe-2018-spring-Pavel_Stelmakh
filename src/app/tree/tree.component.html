<input type="text" [formControl]="pathout" placeholder="path"><br>
<input type="text" [formControl]="pathin" placeholder="search"><br>
<button (click)="search()">search</button>
<button (click)="resetTree()">reset tree</button>
<div [class.tree]="true">
  <app-tree-item *ngFor="let item of items" [item]="item"></app-tree-item>
</div>