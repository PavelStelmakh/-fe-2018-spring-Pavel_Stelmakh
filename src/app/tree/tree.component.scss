@import '../shared/driveAnimation';

@include animation("list", 0, 20vh, rgba(0, 0, 0, 0), rgba(0, 0, 0, 0.7));

:host {
    display: block;
    color: #ccc;

    .tree {
        overflow: auto;
        width: 500px;
        height: 400px;
    }

    button {
        margin: 3px;        
        color: #ccc;
    }
}
