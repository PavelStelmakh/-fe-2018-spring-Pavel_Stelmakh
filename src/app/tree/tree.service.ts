import { Injectable, OnDestroy } from '@angular/core';
import { Subject } from 'rxjs';
import getTreeItems from '../../../shared/getTreeItems';

@Injectable({
  providedIn: 'root'
})
export class TreeService implements OnDestroy {
  items: string[] = getTreeItems;
  path: string[] = [];
  pathupdate: Subject<void> = new Subject<void>();
  search: Subject<string> = new Subject<string>();

  constructor() {}

  ngOnDestroy() {
    this.search.complete();
    this.pathupdate.complete();
  }

  getTreeItems(): string[] {
    return this.items;
  }

  pushPath(item: string) {
    this.path.push(item);
    this.pathupdate.next();
  }

  setPath(index: number, item: string) {
    this.path[index] += `,${item}`;
    this.pathupdate.next();
  }

  popPath(items: string[]) {
    this.path = items.slice();
    this.pathupdate.next();
  }

  resetPath() {
    this.path = [];
    this.pathupdate.next();
  }

}
