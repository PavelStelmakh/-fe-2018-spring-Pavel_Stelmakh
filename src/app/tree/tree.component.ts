import { Component, OnInit, ViewChildren, QueryList } from '@angular/core';
import { TreeService } from './tree.service';
import { FormControl } from '@angular/forms';
import { TreeItemComponent } from './tree-item/tree-item.component';

@Component({
  selector: 'app-tree',
  templateUrl: './tree.component.html',
  styleUrls: ['./tree.component.scss']
})
export class TreeComponent implements OnInit {
  items: string[];
  pathout: FormControl;
  pathin: FormControl;
  @ViewChildren(TreeItemComponent) itemComponents: QueryList<TreeItemComponent>;

  constructor(public treeService: TreeService) { }

  ngOnInit() {
    this.items = this.treeService.getTreeItems();
    this.pathout = new FormControl({value: '', disabled: true});
    this.pathin = new FormControl('');
    this.treeService.pathupdate.subscribe(() => this.pathout.setValue(this.treeService.path.join('->')));
    this.treeService.search.subscribe((path: string) => {
      this.filter(path);
    });
  }

  resetTree() {
    this.items = this.treeService.getTreeItems();
    this.treeService.resetPath();
  }

  search() {
    const searchpath: string = this.pathin.value.trim();
    this.treeService.search.next(searchpath);
  }

  filter(Searchingitem: string) {
    this.items = this.itemComponents
    .filter((component: TreeItemComponent) => Searchingitem === component.itemTree || component.expandItem)
    .map((component: TreeItemComponent) => component.itemTree);
  }

}
