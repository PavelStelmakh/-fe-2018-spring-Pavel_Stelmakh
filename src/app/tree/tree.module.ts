import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FontAwesomeModule } from '@fortawesome/angular-fontawesome';
import { ReactiveFormsModule } from '@angular/forms';
import { TreeComponent } from './tree.component';
import { TreeItemComponent } from './tree-item/tree-item.component';
import { TreeService } from './tree.service';

@NgModule({
  imports: [
    CommonModule,
    FontAwesomeModule,
    ReactiveFormsModule
  ],
  declarations: [TreeComponent, TreeItemComponent],
  providers: [TreeService],
  exports: [TreeComponent]
})
export class TreeModule { }
