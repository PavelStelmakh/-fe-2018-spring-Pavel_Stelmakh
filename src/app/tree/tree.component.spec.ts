import { TestBed, async, ComponentFixture } from '@angular/core/testing';
import { ReactiveFormsModule } from '@angular/forms';
import { NO_ERRORS_SCHEMA } from '@angular/core';
import { TreeComponent } from './tree.component';
import { TreeService } from './tree.service';

describe('TreeComponent', () => {
  let fixture: ComponentFixture<TreeComponent>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      imports: [
        ReactiveFormsModule
      ],
      declarations: [
        TreeComponent
      ],
      providers: [
        TreeService
      ],
      schemas: [
       NO_ERRORS_SCHEMA
      ]
    }).compileComponents();

  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(TreeComponent);
  });

  it('renders markup to snapshot', () => {
    expect(fixture).toMatchSnapshot();
  });

});
