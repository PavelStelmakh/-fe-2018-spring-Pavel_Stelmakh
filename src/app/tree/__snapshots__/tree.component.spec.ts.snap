// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TreeComponent renders markup to snapshot 1`] = `
<app-tree
  treeService={[Function TreeService]}
>
  <input
    placeholder="path"
    type="text"
  /><br /><input
    placeholder="search"
    type="text"
  /><br /><button>
    search
  </button><button>
    reset tree
  </button><div>
    
  </div>
</app-tree>
`;
